File: Difficulty/Medium/[144] Binary Tree Preorder Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        if not root: return []
        res, s = [], []
        s.append(root)
        while s:
            visit = s.pop()
            res.append(visit.val)
            if visit.right:
                s.append(visit.right)
            if visit.left:
                s.append(visit.left)
        return res

File: Difficulty/Medium/[144] Binary Tree Preorder Traversal/solution.cpp
/*
 * @lc app=leetcode id=144 lang=cpp
 *
 * [144] Binary Tree Preorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if (! root) return vector<int>();
        vector<int> res;
        stack<TreeNode*> s;
        s.push(root);
        while (!s.empty()) {
            TreeNode* visit = s.top();
            res.push_back(visit->val);
            s.pop();
            if (visit->right) s.push(visit->right);
            if (visit->left) s.push(visit->left);
        }
        return res;
    }
};
// @lc code=end

